### PR TITLE
obs-filters: Fix NVIDIA greenscreen issues

### PR DESCRIPTION
**Pull request by jpark37**
### Description
Fix non-async sources, make sure math is linear-correct, preserve
incoming alpha, and allow the filter to work in previews.

### Motivation and Context
Improve color correctness.

### How Has This Been Tested?
Verified main preview and filter preview look reasonable.

### Types of changes
- Bug fix (non-breaking change which fixes an issue)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

**Comment by pkviet** [comment]
There was a bug with async sources that rcdrone fixed, all my tests are now fine.
LGTM

**Review by pkviet — APPROVED** [review]
Thanks a lot.
This fixes and improves the filter as advertised.

File: plugins/obs-filters/nvidia-greenscreen-filter.c [hunk 1]
```diff
@@ -248,6 +260,10 @@ static void init_images_greenscreen(struct nv_greenscreen_data *filter)
 	}
 
 	/* 5. Create and allocate Alpha NvCVimage (fx dst). */
+	if (filter->A_dst_img) {
+		NvCVImage_Destroy(filter->A_dst_img);
+		filter->A_dst_img = NULL;
+	}
```
Comment (theHamsta) [inline plugins/obs-filters/nvidia-greenscreen-filter.c:266]: `NvCVImage_Realloc` might be usefull for some of the destroy/alloc combinations

Comment (jpark37) [reply plugins/obs-filters/nvidia-greenscreen-filter.c]: I removed the destroy calls, and tested resize with debugger breakpoints at all 5 spots. I assume NvCVImage_InitFromD3D11Texture has a "realloc" effect.

File: plugins/obs-filters/nvidia-greenscreen-filter.c [hunk 2]
```diff
@@ -517,8 +546,10 @@ static void nv_greenscreen_filter_tick(void *data, float t)
 		filter->height = cy;
 	}
 	if (!filter->images_allocated) {
+		obs_enter_graphics();
 		init_images_greenscreen(filter);
-		filter->initial_render = 0;
+		obs_leave_graphics();
```
[the remark this answers is not in the document]
Comment (jpark37) [reply plugins/obs-filters/nvidia-greenscreen-filter.c:551]: I just don't want to think about this function getting sliced. These video ticks are on the graphics thread anyway, so it's not like we're preventing graphics progress.

[the remark this answers is not in the document]
Comment (pkviet) [reply plugins/obs-filters/nvidia-greenscreen-filter.c:551]: ok works for me

**Review by pkviet — APPROVED** [review]

**Comment by pkviet** [comment]
Additionally line 687 (https://github.com/obsproject/obs-studio/pull/6988/files#diff-ad05eeb2f463bdb7d7bf87899eb00a98844be125267430426603aa6b9bb711e3R687) fixes a regression introduced in https://github.com/obsproject/obs-studio/commit/f1797f2c52e4fe26e86941c33bd12c7f3261ba23 where the fx processing would not work any more for non async parent sources.

**Pull request merged**